feat: Add support for PDBs on deployment and statefulset

JorTurFer opened the pull request:
Fixes: https://github.com/open-telemetry/opentelemetry-operator/issues/2136

JorTurFer commented:
Sorry, I missed the bundle at the first try. I have updated the PR

JorTurFer commented:
awesome, a missing change in the stage 🤦

jaronoff97 commented on controllers/opentelemetrycollector_controller.go:
```diff
@@ -305,11 +308,18 @@ func (r *OpenTelemetryCollectorReconciler) SetupWithManager(mgr ctrl.Manager) er
 
 	autoscalingVersion := r.config.AutoscalingVersion()
-	if autoscalingVersion == autodetect.AutoscalingVersionV2 {
+	if autoscalingVersion == autodetect.VersionV2 {
```
Why is this changing?

jaronoff97 commented on apis/v1alpha1/opentelemetrycollector_types.go:
```diff
@@ -123,6 +124,11 @@ type OpenTelemetryCollectorSpec struct {
+	// for the OpenTelemetryCollector workload.
+	//
+	// +optional
+	PodDisruptionBudget *PodDisruptionBudgetSpec `json:"podDisruptionBudget,omitempty"`
```
note for other reviewers, this is a paired down version of the main PDB spec because we don't want to allow the configuration of the selector

jaronoff97 commented on internal/manifests/collector/poddisruptionbudget.go:
```diff
@@ -0,0 +1,78 @@
+	"github.com/open-telemetry/opentelemetry-operator/pkg/autodetect"
+)
+
+func PodDisruptionBudget(cfg config.Config, logger logr.Logger, otelcol v1alpha1.OpenTelemetryCollector) client.Object {
```
can we add a comment here linking to the deprecation docs so we know when we can remove the dual state? https://kubernetes.io/docs/reference/using-api/deprecation-guide/#poddisruptionbudget-v125

pavolloffay replied:
I am fine bumping the lowest support k8s version to 1.21. 

It seems that the oldest supported k8s version by a cloud provider out there is 1.23. I will start a GH issue to discuss it.

JorTurFer replied:
If the lowest version is 1.21, we don't need to autodetect versions for pdb, and if the lowest version would be 1.23, the same applies for autoscaling (autoscaling/v2 was introduced on 1.23).
I don't have problems to update my PR if you have consensus about the lowest version change

jaronoff97 reviewed (changes requested):
Overall this looks good! Thank you for updating the logic for version detection as well to be more generic. Could you also update one of the e2e to set the PDB and ensure it gets reconciled correctly and the selector works as expected?

JorTurFer commented:
Thanks for the quick review, tomorrow I'll address al the comments 😄

JorTurFer commented:
I have addressed all the comments. I'm not 100% if I have done the e2e test correctly because it's the 1st time I have used kuttl but at least it works in local 🤣

JorTurFer commented:
It looks like I have to review the e2e test to work on multiple cluster versions. Let me do it before triggering again the CI because the check will fail again. I'm on it

JorTurFer commented:
@pavolloffay ,
What should I do? I mean, should I wait until https://github.com/open-telemetry/opentelemetry-operator/issues/2145 is decided? Will it require too much time? I'm not in rush with the pdb, during next month is enough for use, but if the discussion goes longes I'd prefer to merge it with support to both

pavolloffay commented:
> What should I do?

Let's see how quickly goes the conversation in https://github.com/open-telemetry/opentelemetry-operator/issues/2145 I am happy to open a PR next week to bump the version.

JorTurFer commented:
I have already rebased main branch and updated this PR. PTAL when you have some time @pavolloffay

jaronoff97 commented on apis/v1alpha1/opentelemetrycollector_types.go:
```diff
@@ -444,6 +450,23 @@ type AutoscalerSpec struct {
+	// absence of the evicted pod.  So for example you can prevent all voluntary
+	// evictions by specifying "100%".
+	// +optional
+	MinAvailable *intstr.IntOrString `json:"minAvailable,omitempty"`
```
does one of these need to be set in order to be accepted?

JorTurFer replied:
What do you mean?

jaronoff97 replied:
does someone who wants to use a PDB need to set either min or max available for the kube api to accept the object?

JorTurFer replied:
No no, the must set only one of them, they are mutually exclusive. If they set both, the admission webhook will reject the CR: https://github.com/open-telemetry/opentelemetry-operator/pull/2141/files#diff-f7d2d1348f61f254e8b086ada3e9ad6ccc705b9cbbd171065911fd75b38bf495R251-R255.

jaronoff97 replied:
ah i missed it in my initial review, thank you! This LGTM.

JorTurFer replied:
no worries, it wasn't in your initial review, I added it due to your feedback 😄

jaronoff97 reviewed (commented):
I think we may want to add something in the validating webhook to be sure that we don't attempt a bad config

pavolloffay commented:
`End-to-end tests / End-to-end tests (1.23, e2e-pdb) (pull_request) ` - failed

JorTurFer commented:
I'm spinning up a cluster with that version to check it

JorTurFer commented:
@pavolloffay , I've tested it locally with k8s 1.23 and (at least locally) the feature works:
![image](https://github.com/open-telemetry/opentelemetry-operator/assets/36899226/17409699-5fe3-4b6e-8458-49f083773810)

The problem I have seen is that the deletion is producing timeouts

JorTurFer commented:
I've noticed that if I use my own namespace for the tests, the k8s GC removes the namespaces more quickly and I can also validate the generated labels, so I've updated my test to use it's own namespace, and now 30 seconds of timeout is more than enough (the test needs a few seconds).

I say this because I see other e2e tests with > 150 seconds of timeouts and maybe the root cause it's something like this

pavolloffay commented:
The PR looks good to me. But I am also interested in @jaronoff97 question whether we could do some validation of the PDB fields that we added to our CR.

JorTurFer commented:
Which kind of validation? I mean, I added a check in webhooks to validate if both have been set, and the other validations are already done by the openapi spec (but I can be missing any important thing at some point)

jaronoff97 commented:
@JorTurFer thank you so much for your contribution 🎊🎉
